0023-bilateral-transfer-protocol/0023-bilateral-transfer-protocol.md

```diff
@@ -175,6 +180,31 @@ ProtocolData ::= SEQUENCE OF SEQUENCE {
 }
 ```
 
+## Authentication
+
+Before anything else, when a client connects to a server, it sends a special
+`Message` request. Its primary `protocolData` entry MUST have name `'auth'`
+and empty data, and among the secondary entries, there MUST be a UTF-8
+`'auth_token'` entry, and a UTF-8 `'auth_username'` entry. The further secondary
```

Comment:
`auth_username` may not be present in some cases (e.g. peering) where the username is implicit in the endpoint being connected to.

Reply:
That's asymmetric then: `auth_token` should be UTF-8 and either the token or the empty string, and `auth_username` should either be UTF-8 and the username, or missing? I don't see anything in Stefan's notes about that, but if you remember that's how we agreed it on Monday, then that's what I'll write.

Reply:
see also further comments on https://github.com/interledgerjs/ilp-plugin-payment-channel-framework/commit/f3f921a6240971127dd4b168d439adf7a5b8650a#commitcomment-24424484

Reply:
We discussed this in the #ripple-de-lux slack channel, I think the best option is:
* A server should give out BTP URIs of one form, `btp+wss://auth_username:auth_token@domain.com/path`
* If the server, for whatever reason, doesn't care about the `auth_username`, it can set it to `''` in the BTP URI.
* If the server, for whatever reason, doesn't care about the `auth_token`, it can set it to `''` in the BTP URI.
* a client should always send both `auth_username` and `auth_token` exactly as they appeared in the BTP URI (even if that means `''`)

Reply:
sure, I implemented the first version without auth_username because it isn't an admin plugin, but I can just check to make sure its there and empty.